lib/pos_fee_estimator.go

```diff
@@ -94,27 +104,35 @@ func (posFeeEstimator *PoSFeeEstimator) Init(
 func (posFeeEstimator *PoSFeeEstimator) AddBlock(block *MsgDeSoBlock) error {
 	posFeeEstimator.rwLock.Lock()
 	defer posFeeEstimator.rwLock.Unlock()
+
 	if err := posFeeEstimator.addBlockNoLock(block); err != nil {
 		return errors.Wrap(err, "PoSFeeEstimator.AddBlock: error adding block to PoSFeeEstimator")
 	}
+
 	return nil
 }
 
 // addBlockNoLock is the same as AddBlock but assumes the caller has already acquired the rwLock.
 func (posFeeEstimator *PoSFeeEstimator) addBlockNoLock(block *MsgDeSoBlock) error {
+	// Create a new slice to house the new past blocks and add the new block to it.
+	newPastBlocks := append(posFeeEstimator.cachedBlocks, block)
+	newPastBlocks = posFeeEstimator.cleanUpPastBlocks(newPastBlocks)
+
+	// Create a clean transaction register to add the blocks' transactions.
+	newTransactionRegister := NewTransactionRegister()
+	newTransactionRegister.Init(posFeeEstimator.globalParams.Copy())
+
 	// Add all transactions from the block to the pastBlocksTransactionRegister.
-	if err := addBlockToTransactionRegister(posFeeEstimator.pastBlocksTransactionRegister, block); err != nil {
-		return errors.Wrap(err, "PoSFeeEstimator.addBlockNoLock: error adding block to pastBlocksTransactionRegister")
-	}
-	posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks, block)
-	// Sort the cached blocks by height & tstamp just to be safe.
-	posFeeEstimator.sortCachedBlocks()
-	if uint64(len(posFeeEstimator.cachedBlocks)) > posFeeEstimator.numPastBlocks {
-		// Remove the oldest block.
-		if err := posFeeEstimator.removeBlockNoLock(posFeeEstimator.cachedBlocks[0]); err != nil {
-			return errors.Wrap(err, "PoSFeeEstimator.addBlockNoLock: error removing block from PoSFeeEstimator")
+	for _, pastBlock := range newPastBlocks {
```

Comment:
Rather than trying to incrementally manage the `pastBlocksTransactionRegister` as new blocks come in, it ends up being less error prone but just as performant to nuke it and recreate a new one from scratch. The number of past blocks will be small (<= 50) so this is not an issue.